Subject: Linux-RT 2.6.27-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Tony Jones <tonyj@suse.de>
From: Steven Rostedt <srostedt@redhat.com>
Subject: ftrace: avoid lockdep recursion

The function_trace_call needs to find the cpu it is running on to be able
to disable the trace to pretect from recursion. It currently performs a
local_irq_save, but when lockdep is on, this may cause unwanted recursion
if lockdep happens to call something that is traced.

This patch changes this to use preempt_disable instead.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>

---
 kernel/trace/trace.c |   22 ++++++++++++++++++----
 1 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/kernel/trace/trace.c b/kernel/trace/trace.c
index c632066..f88e031 100644
--- a/kernel/trace/trace.c
+++ b/kernel/trace/trace.c
@@ -1029,7 +1029,7 @@ function_trace_call(unsigned long ip, unsigned long parent_ip)
 	struct trace_array_cpu *data;
 	unsigned long flags;
 	long disabled;
-	int cpu;
+	int cpu, resched;
 
 	if (unlikely(!ftrace_function_enabled))
 		return;
@@ -1037,16 +1037,30 @@ function_trace_call(unsigned long ip, unsigned long parent_ip)
 	if (skip_trace(ip))
 		return;
 
-	local_irq_save(flags);
+	resched = need_resched();
+	preempt_disable_notrace();
 	cpu = raw_smp_processor_id();
 	data = tr->data[cpu];
 	disabled = atomic_inc_return(&data->disabled);
 
-	if (likely(disabled == 1))
+	if (likely(disabled == 1)) {
+		local_save_flags(flags);
 		trace_function(tr, data, ip, parent_ip, flags);
+	}
 
 	atomic_dec(&data->disabled);
-	local_irq_restore(flags);
+
+	/*
+	 * To prevent recursion with schedule(), we look at the
+	 * resched flag before disabling preemption. If it is already
+	 * set, then we may be just calling schedule, and we don't
+	 * want to call schedule again. But if it was not set, then
+	 * it is fine to call schedule() if we need to.
+	 */
+	if (resched)
+		preempt_enable_no_resched_notrace();
+	else
+		preempt_enable_notrace();
 }
 
 static struct ftrace_ops trace_ops __read_mostly =
